src/c_api/adapter.h

```diff
@@ -0,0 +1,275 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file adapter.h
+ */
+#ifndef XGBOOST_C_API_ADAPTER_H_
+#define XGBOOST_C_API_ADAPTER_H_
+#include <string>
+#include <limits>
+namespace xgboost {
+
+/** \brief External data formats should implement an adapter as below. The
+ * adapter provides a uniform access to data outside xgboost, allowing
+ * construction of DMatrix objects from a range of sources without duplicating
+ * code. The adapter should translate external data into batches of COO tuples
+ * containing the floating point value, row index and column index.*/
+class ExternalDataAdapter {
+ public:
+  ExternalDataAdapter(size_t num_features, size_t num_rows, size_t num_elements)
+      : num_features(num_features),
+        num_rows(num_rows),
+        num_elements(num_elements) {}
+  struct COOTuple {
+    COOTuple(size_t row_idx, size_t column_idx, float value)
+        : row_idx(row_idx), column_idx(column_idx), value(value) {}
+
+    size_t row_idx{0};
+    size_t column_idx{0};
+    float value{0};
+  };
+  size_t GetNumFeatures() const { return num_features; }
+  size_t GetNumRows() const { return num_rows; }
+  size_t GetNumElements() const { return num_elements; }
+
+ protected:
+  size_t num_features;
+  size_t num_rows;
+  size_t num_elements;
+};
+
+class CSRAdapter : public ExternalDataAdapter {
+ public:
+  CSRAdapter(const size_t* row_ptr, const unsigned* feature_idx,
+             const float* values, size_t num_rows, size_t num_elements,
+             size_t num_features)
+      : ExternalDataAdapter(num_features, num_rows, num_elements),
+        row_ptr(row_ptr),
+        feature_idx(feature_idx),
+        values(values) {}
+
+ private:
+  class CSRAdapterBatch {
```

Comment:
Why do we need an intermedia type `Batch`?

Reply:
We need to access the data in parts on the host when using sparse formats because we cannot efficiently perform random access reads into a sparse matrix. If I tried to randomly access an element in a CSR matrix, the floating point value and column index are immediately available but I need to perform binary search to find the row index. 

So it is efficient to access elements by iterating through the rows of a CSR matrix or the columns of a CSC matrix, but not by random access. 

On the GPU we will simply do the binary search because we have a lot more computing power.